libp2p/security/security_multistream.py

```diff
@@ -0,0 +1,88 @@
+from abc import ABC
+from libp2p.protocol_muxer.multiselect_client import MultiselectClient
+from libp2p.protocol_muxer.multiselect import Multiselect
+
+# pylint: disable=W0105
+
+"""
+Represents a secured connection object, which includes a connection and details about the security
+involved in the secured connection
+
+Relevant go repo: https://github.com/libp2p/go-conn-security/blob/master/interface.go
+"""
+class SecurityMultistream(ABC):
+
+    def __init__(self):
+        # Map protocol to secure transport
+        self.transports = {}
+
+        # Create multiselect
+        self.multiselect = Multiselect()
+
+        # Create multiselect client
+        self.multiselect_client = MultiselectClient()
+
+    def add_transport(self, protocol, transport):
+        # Associate protocol with transport
+        self.transports[protocol] = transport
+
+        # Add protocol and handler to multiselect
+        # Note: None is added as the handler for the given protocol since
+        # we only care about selecting the protocol, not any handler function
+        self.multiselect.add_handler(protocol, None)
+
+
+    async def secure_inbound(self, conn):
+        """
+        Secure the connection, either locally or by communicating with opposing node via conn,
+        for an inbound connection (i.e. we are not the initiator)
+        :return: secure connection object (that implements secure_conn_interface)
+        """
+
+        # Select a secure transport
+        transport = await self.select_transport(conn, False)
+
+        # Create secured connection
+        secure_conn = await transport.secure_inbound(conn)
+
+        return secure_conn
+
+
+    async def secure_outbound(self, conn, peer_id):
+        """
+        Secure the connection, either locally or by communicating with opposing node via conn,
+        for an inbound connection (i.e. we are the initiator)
+        :return: secure connection object (that implements secure_conn_interface)
+        """
+
+        # Select a secure transport
+        transport = await self.select_transport(conn, True)
+
+        # Create secured connection
+        secure_conn = await transport.secure_outbound(conn, peer_id)
+
+        return secure_conn
+
+
+    async def select_transport(self, conn, initiator):
+        """
+        Select a transport that both us and the node on the
+        other end of conn support and agree on
+        :param conn: conn to choose a transport over
+        :param initiator: true if we are the initiator, false otherwise
+        :return: selected secure transport
+        """
+        # TODO: Is conn acceptable to multiselect/multiselect_client
```

Comment:
multiselect takes in a stream in our repo, i can buy the argument of passing in a raw conn since security upgrade should happen before transport upgrade but we may have to refactor some of the multiselect code. To be exact, MultistreamCommunicator expects a stream as input but we are not treating it any differently than a raw connection in our code.

Reply:
We don’t need to treat it different because raw conn has the same read and write functions as stream now and in python if it walks like a duck and talks like a duck it’s a duck

Reply:
I know, but the naming in our code can be confusing to future contributors.

Reply:
So what are you saying we should change?

Reply:
this could work as a short-term fix but i feel like if it doesnt need to be a stream, we should probably name it something else. I dont have any better idea off the top of my head.

Reply:
True, we could have an interface with a `read` and `write` function.

Reply:
Added the comment in this PR. If we want to refactor that could be done in another PR since there's other classes that we may want to implement that sort of reader+writer interface